server/lib/chat/chat-service.ts

```diff
@@ -174,6 +181,90 @@ export default class ChatService {
     this.unsubscribeUserFromChannel(userSockets, originalChannelName)
   }
 
+  async moderateUser(
+    channelName: string,
+    moderatorId: SbUserId,
+    targetId: SbUserId,
+    moderationAction: ChannelModerationAction,
+    moderationReason?: string,
+  ): Promise<void> {
+    const [moderatorPermissions, moderatorUserChannel, targetUserChannel] = await Promise.all([
+      getPermissions(moderatorId),
+      getJoinedChannelForUser(moderatorId, channelName),
+      getJoinedChannelForUser(targetId, channelName),
+    ])
+    if (!moderatorUserChannel) {
+      throw new ChatServiceError(
+        ChatServiceErrorCode.InvalidModerationAction,
+        'Must be in channel to moderate users',
+      )
+    }
+    if (!targetUserChannel) {
+      throw new ChatServiceError(
+        ChatServiceErrorCode.InvalidModerationAction,
+        'User must be in channel to moderate them',
+      )
+    }
+    // TODO(2Pac): Introduce a concept of channel "owner" who can't be moderated
+    if (
+      !moderatorPermissions.moderateChatChannels &&
+      !moderatorUserChannel.channelPermissions.editPermissions
+    ) {
+      if (!moderatorUserChannel.channelPermissions[moderationAction]) {
+        throw new ChatServiceError(
+          ChatServiceErrorCode.ModeratorAccess,
+          'Not enough permissions to moderate the user',
+        )
+      }
+      if (
+        targetUserChannel.channelPermissions.editPermissions ||
+        targetUserChannel.channelPermissions.ban ||
+        targetUserChannel.channelPermissions.kick
+      ) {
+        throw new ChatServiceError(
+          ChatServiceErrorCode.ModeratorAccess,
+          "Can't moderate users with moderator access",
+        )
+      }
+    }
+
+    const originalChannelName = await this.getOriginalChannelName(channelName)
+    if (originalChannelName === 'ShieldBattery') {
+      throw new ChatServiceError(
+        ChatServiceErrorCode.LeaveShieldBattery,
+        "Can't moderate users in the ShieldBattery channel",
+      )
+    }
+    if (
+      !this.state.users.has(targetId) ||
+      !this.state.users.get(targetId)!.has(originalChannelName)
+    ) {
+      throw new ChatServiceError(
+        ChatServiceErrorCode.InvalidModerationAction,
+        'User must be in channel to moderate them',
+      )
+    }
+
+    const targetSockets = this.getUserSockets(targetId)
+    // NOTE(2Pac): New owner can technically be selected if a global moderator removes all users
+    // with channel permissions.
+    const newOwner = await this.removeUserFromChannel(originalChannelName, targetSockets)
```

Comment:
this should really be done in a way that notifies the user they've been kicked, shouldn't be handled in a way where that is ambiguous to leaving (because the user may have multiple clients connected, so the clients really can't tell either way)